src/core/datatypes/math/OrderedEvents.h

```diff
@@ -0,0 +1,232 @@
+/*
+ * EventTime.h
+ *
+ *  Created on: Apr 9, 2020
+ *      Author: mike
+ */
+
+#ifndef ORDEREDEVENTS_H_
+#define ORDEREDEVENTS_H_
+
+#include <map>
+
+#include "Cloneable.h"
+
+namespace RevBayesCore {
+
+	template<class valueType>
+	class OrderedEvents : public Cloneable, public MemberObject< RbVector<valueType> > {
+
+	public:
+
+		                                   OrderedEvents();
+		virtual                           ~OrderedEvents();
+
+		OrderedEvents*                     clone(void) const;
+
+        bool                               operator==(const OrderedEvents &oe) const;
+        bool                               operator!=(const OrderedEvents &oe) const;
+        bool                               operator<(const  OrderedEvents &oe) const;
+        bool                               operator<=(const OrderedEvents &oe) const { return operator<(oe) || operator==(oe); }
+
+		// getters and setters
+		bool                               addEvent(double time, valueType event);
+		bool                               removeEvent(double time);
+		bool                               changeEventTime(double old_time, double new_time);
+		bool                               changeEvent(double time, valueType new_value);
+		const std::map<double, valueType>& getEvents() const;
+		size_t                             size() const;
+
+		// exposed methods
+        void                               executeMethod(const std::string &n, const std::vector<const DagNode*> &args, RbVector<valueType> &rv) const; //!< Map the member methods to internal function calls
+
+	private:
+
+		std::map<double, valueType>  events;
```

Comment:
You could in theory use a std::multimap<double, valueType>, which would allow two events to have the same time.
I assume the idea is that events occurring at random times will never have exactly the same time?